MP OpenAPI: Warn if an MP Config property is ignored because equivalent or conflicting config is provided in server.xml

Azquelt opened the pull request:
MP OpenAPI 2.0 and later allow certain configuration options to be provided via server.xml or MP Config. If config is provided in server.xml, then any config from MP Config is ignored.

This PR adds a message when an MP Config property is ignored. It's a warning if the config in MP Config is different and an info message if it is equivalent.

Also does some further refactoring of `ModuleSelectionConfigImpl`.

Fixes #30046 

Also fixes the `equals` method in `MpConfigServerConfigObject` which in some circumstances could cause the server to reprocess applications unnecessarily.

Fixes #30340

Also adds the message for #30036 since I thought the translation deadline was sooner than it actually is.

- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Azquelt committed:
Add new MP OpenAPI warning/info messages

Adding these before the implementation to get them in for the
translation deadline.

Co-authored-by: Ramkumar <88021034+ramkumar-k-9286@users.noreply.github.com>

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -161,3 +161,15 @@ OPENAPI_INFO_INVALID_SERVERXML_CWWKO1683W.useraction=Ensure that the title and v
 OPENAPI_MERGE_UNUSED_INCLUDE_SERVERXML_CWWKO1684W.explanation=The server is configured to create OpenAPI documentation for the application or module name, but no application or web module with that name was deployed.
 OPENAPI_MERGE_UNUSED_INCLUDE_SERVERXML_CWWKO1684W.useraction=Check that all applications started correctly and that the names in the configuration are specified correctly.
+
+OPENAPI_MP_CONFIG_REDUNDANT_CWWKO1685I=CWWKO1685I: The {0} MicroProfile Config property is ignored because equivalent configuration is provided in server.xml.
```
```suggestion
OPENAPI_MP_CONFIG_REDUNDANT_CWWKO1685I=CWWKO1685I: The {0} MicroProfile Config property is ignored because the equivalent configuration is provided in the server.xml file.
```

Azquelt replied:
I don't think "because the equivalent configuration" is correct here. I think this statement has the same structure as the message further down the file where we have "because conflicting configuration".

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -161,3 +161,15 @@ OPENAPI_INFO_INVALID_SERVERXML_CWWKO1683W.useraction=Ensure that the title and v
 OPENAPI_MERGE_UNUSED_INCLUDE_SERVERXML_CWWKO1684W.useraction=Check that all applications started correctly and that the names in the configuration are specified correctly.
+
+OPENAPI_MP_CONFIG_REDUNDANT_CWWKO1685I=CWWKO1685I: The {0} MicroProfile Config property is ignored because equivalent configuration is provided in server.xml.
+OPENAPI_MP_CONFIG_REDUNDANT_CWWKO1685I.explanation=When this property is set and the same function is configured using server.xml, the configuration in server.xml is used and the MicroProfile Config property is ignored.
```
```suggestion
OPENAPI_MP_CONFIG_REDUNDANT_CWWKO1685I.explanation=When this property is set and the same function is configured by using the server.xml file, the configuration in the server.xml file is used and the MicroProfile Config property is ignored.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -161,3 +161,15 @@ OPENAPI_INFO_INVALID_SERVERXML_CWWKO1683W.useraction=Ensure that the title and v
+OPENAPI_MP_CONFIG_REDUNDANT_CWWKO1685I.useraction=Remove the MicroProfile Config property.
+
+OPENAPI_MP_CONFIG_CONFLICTS_CWWKO1686W=CWWKO1686W: The {0} MicroProfile Config property is ignored because conflicting configuration is provided in server.xml.
+OPENAPI_MP_CONFIG_CONFLICTS_CWWKO1686W.explanation=When this property is set and the same function is configured using server.xml, the configuration in server.xml is used and the MicroProfile Config property is ignored.
```
```suggestion
OPENAPI_MP_CONFIG_CONFLICTS_CWWKO1686W.explanation=When this property is set and the same function is configured by using the server.xml file, the configuration in the server.xml file is used and the MicroProfile Config property is ignored.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -161,3 +161,15 @@ OPENAPI_INFO_INVALID_SERVERXML_CWWKO1683W.useraction=Ensure that the title and v
+
+OPENAPI_MP_CONFIG_CONFLICTS_CWWKO1686W=CWWKO1686W: The {0} MicroProfile Config property is ignored because conflicting configuration is provided in server.xml.
+OPENAPI_MP_CONFIG_CONFLICTS_CWWKO1686W.explanation=When this property is set and the same function is configured using server.xml, the configuration in server.xml is used and the MicroProfile Config property is ignored.
+OPENAPI_MP_CONFIG_CONFLICTS_CWWKO1686W.useraction=Remove the MicroProfile Config property and ensure that the configuration in server.xml is correct.
```
```suggestion
OPENAPI_MP_CONFIG_CONFLICTS_CWWKO1686W.useraction=Remove the MicroProfile Config property and make sure that the configuration in the server.xml file is correct.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -161,3 +161,15 @@ OPENAPI_INFO_INVALID_SERVERXML_CWWKO1683W.useraction=Ensure that the title and v
+OPENAPI_MP_CONFIG_REDUNDANT_CWWKO1685I.explanation=When this property is set and the same function is configured using server.xml, the configuration in server.xml is used and the MicroProfile Config property is ignored.
+OPENAPI_MP_CONFIG_REDUNDANT_CWWKO1685I.useraction=Remove the MicroProfile Config property.
+
+OPENAPI_MP_CONFIG_CONFLICTS_CWWKO1686W=CWWKO1686W: The {0} MicroProfile Config property is ignored because conflicting configuration is provided in server.xml.
```
```suggestion
OPENAPI_MP_CONFIG_CONFLICTS_CWWKO1686W=CWWKO1686W: The {0} MicroProfile Config property is ignored because conflicting configuration is provided in the server.xml file.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -161,3 +161,15 @@ OPENAPI_INFO_INVALID_SERVERXML_CWWKO1683W.useraction=Ensure that the title and v
+OPENAPI_MP_CONFIG_CONFLICTS_CWWKO1686W.useraction=Remove the MicroProfile Config property and ensure that the configuration in server.xml is correct.
+
+OPENAPI_ANNOTATION_TOO_NEW_CWWKO1687W=CWWKO1687W: The {0} annotation element in the {1} class is ignored because it requires OpenAPI version {2} but the current OpenAPI version is {3}.
+OPENAPI_ANNOTATION_TOO_NEW_CWWKO1687W.explanation=The annotation element relates to a feature which was added to a newer version of the OpenAPI specification than the one that is configured. The OpenAPI version in use does not have a way to represent this information so it will not be included in the produced OpenAPI documentation.
```
```suggestion
OPENAPI_ANNOTATION_TOO_NEW_CWWKO1687W.explanation=The annotation element relates to a feature added to a newer version of the OpenAPI specification than the one configured. The OpenAPI version in use cannot represent this information, so it is excluded from the produced OpenAPI documentation.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -161,3 +161,15 @@ OPENAPI_INFO_INVALID_SERVERXML_CWWKO1683W.useraction=Ensure that the title and v
+
+OPENAPI_ANNOTATION_TOO_NEW_CWWKO1687W=CWWKO1687W: The {0} annotation element in the {1} class is ignored because it requires OpenAPI version {2} but the current OpenAPI version is {3}.
+OPENAPI_ANNOTATION_TOO_NEW_CWWKO1687W.explanation=The annotation element relates to a feature which was added to a newer version of the OpenAPI specification than the one that is configured. The OpenAPI version in use does not have a way to represent this information so it will not be included in the produced OpenAPI documentation.
+OPENAPI_ANNOTATION_TOO_NEW_CWWKO1687W.useraction=Consider removing the annotation element or setting openAPIVersion to a version which supports it in server.xml.
```
```suggestion
OPENAPI_ANNOTATION_TOO_NEW_CWWKO1687W.useraction=Consider removing the annotation element or setting the openAPIVersion configuration element to a version which supports it in the server.xml file.
```

ramkumar-k-9286 replied:
is it `openAPIVersion` or  openAPI version ?

Azquelt replied:
It's `openAPIVersion`. I think maybe this message should say "the openAPIVersion configuration element".

Azquelt commented:
@ramkumar-k-9286 Thank you for your review. I've accepted all of your changes and followed up with additional changes mentioned in the comments.

Azquelt committed:
MP OpenAPI: Warn if MP Config properties ignored

Warn if MP Config properties for module inclusion and exclusion or for
Info config are ignored because configuration is provided in the
server.xml file.

Use an appropriate message depending on whether the configuration in MP
Config is equivalent to or conflicts with the configuration in the
server.xml file.

Previously we only read the Info from MP Config when we needed to. Now
we need to check if we need to warn about MP Config being ignored after
we successfully read info config from server.xml.

Adding these warnings for the Info config requires refactoring the MP
Config reading code, so also move it into OpenAPIInfoConfig so that all
the Info reading code is in one place.

Azquelt committed:
MP OpenAPI: refactor ModuleSelectionConfigImpl

Switch from using multiple booleans to an enum to represent special
mutually exclusive inclusion configurations.

Azquelt committed:
MP OpenAPI: Avoid re-getting config

The matches method may be called multiple times per module start and it
was re-reading the config every time. Pass the config in once so that
we are sure we're using a consistent config throughout.

Azquelt committed:
MP OpenAPI: Remove DefaultInclusion enum

InclusionMode is a superset of DefaultInclusion, so we can use
InclusionMode everywhere.

Azquelt committed:
Add std methods to MpConfigServerConfigObject

- hashCode: should always be implemented if equals is implemented
- toString: makes the trace more helpful

Azquelt committed:
Fix MpConfigServerConfigObject.equals

Azquelt commented:
!build
spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi.1.1_fat_tck,com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.microprofile.openapi_fat_tck,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.2.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.1.internal_fat_tck,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat_tck,io.openliberty.microprofile.openapi.ui.internal_fat
